### DISCUSSION TITLE
Pyright complains incorrectly about object not being iterable.

### DISCUSSION THREAD

**Discussion by fkhawajagh** [microsoft/pyright#9134] [Q&A] [answered]
So I came across this 'not a bug' ticket. https://github.com/microsoft/pyright/issues/7916.  I seem to be having a similar issue even when I'm using guards (both positive and negative)

![image](https://github.com/user-attachments/assets/c43ebc57-06a7-484a-bc0b-4c1ab79b6efa)

If I add a default value for the `get` function, the alert goes away.

![image](https://github.com/user-attachments/assets/646fe930-5615-4a9c-b6c3-9350becbae51)

Any thoughts on what I'm doing wrong here?

    "node_modules/pyright": {
      "version": "1.1.383",
      "resolved": "https://registry.npmjs.org/pyright/-/pyright-1.1.383.tgz",
      "integrity": "sha512-b540vUDWGXFlVwhxREgCrvKYT9bnUUPiDtSv5s7sUGxIokTxc06bPC2vfnGunUqaUu6hgIqlv1GRFdOKIEb09A==",

**Comment by erictraut** [comment]
I'm happy to help you diagnose the problem, but could you please post a minimal repro in text form so I don't need to retype your code from the screen shots?

**Reply by fkhawajagh** [reply]
Thank you .. here's the code

```
def some_func():
    x = {
        'data': [
            {'id': 1, 'name': 'test1'},
            {'id': 2, 'name': 'test2'},
            {'id': 3, 'name': 'test3'},
            {'id': 4, 'name': 'test4'}
        ],
        'other': []
    }

    if x.get('data'):
        for row in x.get('data', None):
            print(f'id: {row.get("id")}')
```

**Reply by erictraut** [reply, marked as answer]
Thanks for the code.

You didn't state this, but I presume that you're using strict type checking mode or have manually enabled the `strictDictionaryInference` configuration option. Without this, I'm not able to repro the issue.

The problem in this code is that you're assuming that a call to `x.get("data")` implies that a subsequent call to `x.get("data", None")` will never return `None`. A type checker cannot make such an assumption. It must assume that each call is independent of other calls, since functions in Python are not guaranteed to return the same value even if the same arguments are passed to a function multiple times.

Here's how to make this code type safe:

```python
    data = x.get("data")
    if data:
        for row in data:
            print(f'id: {row.get("id")}')
```

Or if you prefer to use a walrus operator:

```python
    if data := x.get("data"):
        for row in data:
            print(f'id: {row.get("id")}')
```

**Reply by fkhawajagh** [reply]
Thanks for the response.  I haven't modified pyright at all.  This is the default config as installed by Mason.  That said, I tried this code (below) which also silences the alert.   I don't like the assignment to `data` because it's the only place in my real function where this object is being referenced and it seemed unnecessary to create an intermediate variable.

```
def some_func():
    x = {
        'data': [
            {'id': 1, 'name': 'test1'},
            {'id': 2, 'name': 'test2'},
            {'id': 3, 'name': 'test3'},
            {'id': 4, 'name': 'test4'}
        ],
        'other': []
    }

    for row in x.get('data', []):
        print(f'id: {row.get("id")}')
```

So the question now is.. if `None` is not allowed, why does this not generate an alert?

```
    for row in x.get('data', None):
        print(f'id: {row.get("id")}')
```

**Reply by erictraut** [reply]
I'm not familiar with mason or how it configures pyright. You may need to reach out to its maintainers for additional details. Perhaps you have a `pyproject.toml` file with a `[tools.pyright]` section?

By default, pyright will not generate an error in your initial code sequence. You can see this [in the pyright playground](https://pyright-play.net/?code=CYUwZgBAzg9gtiA%2BmArgOwMYAoCUAuAKAmIgA8IBeCAbyJPoHJgBDAF2YbwgG077%2BaDAJbBOEAIwAaCAzTMEYhqxBRW4hgF9JfASWrDRXAEzTZ8kIuWqjm7bv76RYgMym5CrkpWtntnfcdDCAAWN3NLb2DNfxIAXTtdBhhWAAsQACcxblidDQIdIUhSADoAcxBWLCY2DnwY4jAYdIh0mAB3CCE0MjKKqpZ2BmkAORg0EDr7egAHdK7KsAMuala23sqAIhENnA0GHAIgA). That's because pyright's default inference rules for dictionary expressions infer a type of `dict[str, Unknown]` for `x` in this case. If you enable strict mode, the type of `x` will be inferred as `dict[str, list[dict[str, int | str]]]`, and you will now see a [type violation error](https://pyright-play.net/?strict=true&code=CYUwZgBAzg9gtiA%2BmArgOwMYAoCUAuAKAmIgA8IBeCAbyJPoHJgBDAF2YbwgG077%2BaDAJbBOEAIwAaCAzTMEYhqxBRW4hgF9JfASWrDRXAEzTZ8kIuWqjm7bv76RYgMym5CrkpWtntnfcdDCAAWN3NLb2DNfxIAXTtdBhhWAAsQACcxblidDQIdIUhSADoAcxBWLCY2DnwY4jAYdIh0mAB3CCE0MjKKqpZ2BmkAORg0EDr7egAHdK7KsAMuala23sqAIhENnA0GHAIgA).

You should see the same behaviors (with and without strict mode enabled) in your last code sample.
